File: 02/solution.hs
import qualified Data.Map.Strict as Map
import Data.List

replaceComma :: Char -> Char
replaceComma c
    | c == ',' = ' '
    | otherwise = c

getMemoryMap :: String -> Map.Map Integer Integer
getMemoryMap input = do
    let values = map read $ words $ map replaceComma input

    let doFold (index, memory) element = (index + 1, Map.insert index element memory)

    snd $ foldl doFold (0, Map.empty) values

solve1 :: IO Integer
solve1 = do
    input <- readFile "./input.txt"
    let memory = getMemoryMap input
    pure $ doSolve memory (memory Map.! 0) 0

doSolve :: Map.Map Integer Integer -> Integer -> Integer -> Integer
doSolve memory 99 _ = memory Map.! 0

doSolve memory opcode index = do
    let op1 = memory Map.! (memory Map.! (index + 1))
    let op2 = memory Map.! (memory Map.! (index + 2))
    
    let result = case opcode of
            1 ->
                op1 + op2
            2 ->
                op1 * op2
            _ ->
                error $ "Index " ++ show index ++ " Opcode " ++ show opcode ++ " Memory " ++ show memory

    let memory2 = Map.insert (memory Map.! (index + 3)) result memory

    doSolve memory2 (memory2 Map.! (index + 4)) (index + 4)


solve2 :: IO (Maybe (Integer, Integer))
solve2 = do
    input <- readFile "./input.txt"
    let memory = getMemoryMap input
    let pairs = [(noun, verb) | noun <- [0..99], verb <- [0..99]]

    pure $ find ((19690720==) . (testPair memory)) pairs

testPair :: Map.Map Integer Integer -> (Integer, Integer) -> Integer
testPair memory (noun, verb) = do
    let memory2 = Map.insert 1 noun $ Map.insert 2 verb memory

    doSolve memory2 (memory2 Map.! 0) 0

File: 02/solution.exs
defmodule Solution do
  # 2692315
  def solve1(file \\ "./input.txt") do
    data = get_data(file)

    do_solve1(data, Map.fetch!(data, 0), 0)
    |> Map.fetch!(0)
  end

  defp do_solve1(data, 99, _index) do
    data
  end

  defp do_solve1(data, opcode, index) when opcode in [1, 2] do
    op1 = Map.fetch!(data, Map.fetch!(data, index + 1))
    op2 = Map.fetch!(data, Map.fetch!(data, index + 2))

    result =
      case opcode do
        1 ->
          op1 + op2

        2 ->
          op1 * op2
      end

    data = Map.put(data, Map.fetch!(data, index + 3), result)

    do_solve1(data, Map.fetch!(data, index + 4), index + 4)
  end

  # Reads the file and parses it into a usable format
  defp get_data(file) do
    file
    |> File.read!()
    |> String.trim_trailing("\n")
    |> String.split(",")
    |> Enum.with_index()
    # Use a map to address memory efficiently
    |> Map.new(fn {e, i} -> {i, String.to_integer(e)} end)
  end

  # {95, 7}
  # 9507
  def solve2(file \\ "./input.txt") do
    data = get_data(file)

    for noun <- 0..99, verb <- 0..99, reduce: nil do
      # Only try as long we don't have an answer
      nil ->
        data =
          data
          |> Map.put(1, noun)
          |> Map.put(2, verb)

        case do_solve1(data, Map.fetch!(data, 0), 0) |> Map.fetch!(0) do
          19_690_720 ->
            # {noun, verb}
            100 * noun + verb

          _other ->
            nil
        end

      # Otherwise noop
      other ->
        other
    end
  end
end
